File: typescript/food-chain/food-chain.ts
const LAST_LINE =
  "I don't know why she swallowed the fly. Perhaps she'll die.\n";

interface ChainLink {
  byline?: string;
  next?: Food;
  nextLabel?: string;
}

type Food = | "fly" | "spider" | "bird" | "cat" | "dog" | "goat" | "cow" | "horse";

const CHAIN: Map<Food, ChainLink> = new Map([
  ["fly", {}],
  [
    "spider",
    { byline: "It wriggled and jiggled and tickled inside her.", next: "fly" },
  ],
  [
    "bird",
    {
      byline: "How absurd to swallow a bird!",
      next: "spider",
      nextLabel: "spider that wriggled and jiggled and tickled inside her",
    },
  ],
  ["cat", { byline: "Imagine that, to swallow a cat!", next: "bird" }],
  ["dog", { byline: "What a hog, to swallow a dog!", next: "cat" }],
  [
    "goat",
    { byline: "Just opened her throat and swallowed a goat!", next: "dog" },
  ],
  ["cow", { byline: "I don't know how she swallowed a cow!", next: "goat" }],
  ["horse", { byline: "She's dead, of course!" }],
]);

export default class FoodChain {
  private static unwindFoodChain(food: Food, chain: string): string {
    let link = CHAIN.get(food);
    let nextFood = link?.next;
    let nextFoodLabel = link?.nextLabel || nextFood;

    if (nextFood) {
      let foodChain = FoodChain.unwindFoodChain(nextFood, chain);
      return `She swallowed the ${food} to catch the ${nextFoodLabel}.\n${foodChain}`;
    } else {
      return "";
    }
  }

  private static createVerse(food: Food): string {
    let link = CHAIN.get(food);
    let byline = link?.byline ? "\n" + link?.byline + "\n" : "\n";
    let foodChain = FoodChain.unwindFoodChain(food, "");

    return `I know an old lady who swallowed a ${food}.${byline}${foodChain}`;
  }

  static VERSES: string[] = [
    FoodChain.createVerse("fly") + LAST_LINE,
    FoodChain.createVerse("spider") + LAST_LINE,
    FoodChain.createVerse("bird") + LAST_LINE,
    FoodChain.createVerse("cat") + LAST_LINE,
    FoodChain.createVerse("dog") + LAST_LINE,
    FoodChain.createVerse("goat") + LAST_LINE,
    FoodChain.createVerse("cow") + LAST_LINE,
    FoodChain.createVerse("horse"),
  ];

  static verse(i: number): string {
    return this.VERSES[i - 1];
  }

  static verses(from: number, to: number): string {
    return this.VERSES.slice(from - 1, to).join("\n");
  }
}

File: elixir/food-chain/lib/food_chain.ex
defmodule Food do
  defstruct [:food, :description, :subtitle, :preceeds]
end

defmodule FoodChain do
  @chain [
    %Food{
      food: "fly",
      preceeds: "spider",
      subtitle: "I don't know why she swallowed the fly. Perhaps she'll die."
    },
    %Food{
      food: "spider",
      preceeds: "bird",
      description: " that wriggled and jiggled and tickled inside her",
      subtitle: "It wriggled and jiggled and tickled inside her."
    },
    %Food{
      food: "bird",
      preceeds: "cat",
      subtitle: "How absurd to swallow a bird!"
    },
    %Food{
      food: "cat",
      preceeds: "dog",
      subtitle: "Imagine that, to swallow a cat!"
    },
    %Food{
      food: "dog",
      preceeds: "goat",
      subtitle: "What a hog, to swallow a dog!"
    },
    %Food{
      food: "goat",
      preceeds: "cow",
      subtitle: "Just opened her throat and swallowed a goat!"
    },
    %Food{
      food: "cow",
      preceeds: "horse",
      subtitle: "I don't know how she swallowed a cow!"
    },
    %Food{
      food: "horse",
      preceeds: nil,
      subtitle: "She's dead, of course!"
    }
  ]
  @doc """
  Generate consecutive verses of the song 'I Know an Old Lady Who Swallowed a Fly'.
  """
  @spec recite(start :: integer, stop :: integer) :: String.t()
  def recite(start, start) when start in [1, 8] do
    meal = Enum.at(@chain, start - 1)

    """
    I know an old lady who swallowed a #{meal.food}.
    #{meal.subtitle}
    """
  end

  def recite(start, start) do
    meal = Enum.at(@chain, start - 1)

    """
    I know an old lady who swallowed a #{meal.food}.
    #{meal.subtitle}
    #{add(start - 1)}
    """
    |> String.trim_trailing()
    |> Kernel.<>("\n")
  end

  def recite(start, stop) do
    """
    #{recite(start, start)}
    #{recite(start + 1, stop)}
    """
    |> String.trim_trailing()
    |> Kernel.<>("\n")
  end

  defp add(1) do
    meal = Enum.at(@chain, 0)

    """
    She swallowed the #{meal.preceeds} to catch the #{meal.food}.
    #{meal.subtitle}
    """
  end

  defp add(start) do
    meal = Enum.at(@chain, start - 1)

    """
    She swallowed the #{meal.preceeds} to catch the #{meal.food}#{meal.description}.
    #{add(start - 1)}
    """
  end
end
